google/cloud/storage/async_reader.h

```diff
@@ -0,0 +1,94 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_ASYNC_READER_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_ASYNC_READER_H
+
+#include "google/cloud/storage/async_object_responses.h"
+#include "google/cloud/storage/async_reader_connection.h"
+#include "google/cloud/storage/async_token.h"
+#include "google/cloud/status_or.h"
+#include <memory>
+#include <utility>
+
+namespace google {
+namespace cloud {
+namespace storage_experimental {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+/**
+ * A handle for streaming downloads.
+ *
+ * Application use this object to handle asynchronous streaming downloads. The
+ * application repeatedly calls `Read()` until it has received all the data it
+ * wants.
+ *
+ * The destructor of this object
+ */
+class AsyncReader {
+ public:
+  /// Creates a reader that always returns errors.
+  AsyncReader() = default;
+
+  /// Initialize a reader from its implementation class.
+  explicit AsyncReader(std::unique_ptr<AsyncReaderConnection> impl)
+      : impl_(std::move(impl)) {}
+
+  ///@{
+  /// @name Move-only.
+  AsyncReader(AsyncReader&&) noexcept = default;
+  AsyncReader& operator=(AsyncReader&&) noexcept = default;
+  AsyncReader(AsyncReader const&) = delete;
+  AsyncReader& operator=(AsyncReader const&) = delete;
+  ///@}
+
+  /// Destructor.
+  ~AsyncReader();
+
+  /**
+   * Retrieves more data from the object.
+   *
+   * The returned future becomes satisfied when more data is available. Reading
+   * an object can fail even after the download starts, thus this function
+   * returns wraps the value with `StatusOr<>`. A successful end-of-stream is
+   * indicated by an invalid token.
+   *
+   * @par Example (with coroutines)
+   * @code
+   * future<std::int64_t> CountLines(AsyncReader reader, AsyncToken token) {
+   *   std::int64_t result = 0;
+   *   while (token.valid()) {
+   *     // await for next chunk and throw on error.
```

Comment:
s/await for/wait for/

Reply:
Done. Though I liked the `await` / `co_await` match.